### PR TITLE
Initial NetworkPolicies for System Components

### PR DESCRIPTION
**Pull request by tcdowney**
This is a followup to the ["Default-Deny Network Traffic for System Components"](https://github.com/cloudfoundry/cf-for-k8s/issues/262) we logged the other day.

This PR includes some `NetworkPolicies` that lock deny all egress/ingress for the system components by default and have specific allow policies for known ingress/egress networking. The following diagram shows the interactions we mapped out and discovered through trial and error. It is **not exhaustive** however. We will need help from other component teams to fill in the gaps!

**Update:** The following diagram is now out of date:
system-component-network-interactions

## What We Need
We, the cf-k8s-networking team, don't and can't know all the possible interactions between components. We need the teams that own these components to read over these policies and correct any mistakes that we've made on this initial attempt.

We also need teams to supply information on what **ports** and **protocols** their components use. As written these policies allow any port/protocol and we want to be explicit here.

**It would be awesome if you could just add additional commits to this branch to add these on your policies!**
https://kubernetes.io/docs/concepts/services-networking/network-policies/#networkpolicy-resource
 
## Issues
* The system component namespaces (`cf-system`, `cf-db`, `cf-blobstore`, `cf-workloads-staging`, etc.)  were lacking labels for us to reference. The same issue as https://github.com/cloudfoundry/cf-for-k8s/issues/261. We added some placeholder `name: <namespace-name>` labels in the meantime.
* All pods needed to be able to communicate with platform dependencies running in `kube-system` (such as kube-dns) so we added blanket allow rules for ingress/egress traffic related to that namespace. However, we had to label the namespace ourselves with `name: kube-system` and I'm not sure how something like this would work out of the box. Looking for input here.
* Components that needed to talk to the Kubernetes API (e.g. Eirini, metric-proxy, cf-api-kpack-watcher) or external services (e.g. Cloud Controller, UAA) were given wide-open egress policies. The [external services (identity providers, db, service brokers, etc.) concern](https://github.com/cloudfoundry/cf-for-k8s/issues/262#issuecomment-652056293) deserves its own story/focused effort. Same with the ability to talk to the Kubernetes API -- particularly "off-platform" control planes like on GKE (this is easy if your K8s API lives in `kube-system`).

## Other Thoughts
Writing this YAML by hand was pretty toil-heavy. @XanderStrike had an interesting idea where one might use `ytt` to encode structures representing each arrow on the graph. Like this:

```
#@data/values

---
connections:
- name: migrate-to-db
  fromNs: cf-system
  fromSelector:
    job-name: ccdb-migrate
  toNs: cf-db
  toSelector:
    app: postgresql
- name: api-worker-to-db
  fromNs: cf-system
  fromSelector:
    app.kubernetes.io/name: cf-api-worker
  toNs: cf-db
  toSelector:
    app: postgresql
```

And use ` ytt` loop to generate the `NetworkPolicy` based off of that.
```
#@ load("@ytt:data", "data")

#@ for conn in data.values.connections:
---
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: #@ "{}-egress".format(conn.name)
  namespace: #@ conn.fromNs
spec:
  podSelector:
    matchLabels: #@ conn.fromSelector
  policyTypes:
  - Egress
  egress:
  - to:
    - namespaceSelector:
        matchLabels:
          name:  #@ conn.toNs
      podSelector:
        matchLabels: #@ conn.toSelector
---
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: #@ "{}-ingress".format(conn.name)
  namespace: #@ conn.toNs
spec:
  podSelector:
    matchLabels: #@ conn.toSelector
  policyTypes:
  - Ingress
  ingress:
  - from:
    - namespaceSelector:
         matchLabels:
           name: #@ conn.fromNs
      podSelector:
        matchLabels: #@ conn.fromSelector
```

For the sake of this initial PR I wanted to keep the `NetworkPolicy` as close to how they're actually represented as possible, but would welcome feedback on this alternative approach. We want to make sure we end up with something that fits with the existing `cf-for-k8s` patterns here and is maintainable.

---

Thank you!
Tim && @cloudfoundry/cf-networking 

Tagging some component teams:
cc: @cloudfoundry/cf-capi @cloudfoundry/eirini @cloudfoundry/cf-loggregator @cloudfoundry/cf-metric-store

### CODE REVIEW

File: config/network-policies.yaml [hunk 1]
```diff
@@ -0,0 +1,566 @@
...
+  name: fluentd
+  namespace: #@ data.values.system_namespace
+spec:
+  policyTypes:
+  - Egress
+  podSelector:
+    matchLabels:
+      app: fluentd
+  egress:
+  - {}
```
Comment (tcdowney) [inline config/network-policies.yaml:566]: This is wide-open. Originally I had `fluentd` only allowed to egress to `log-cache`, but it was trying to make an `http` call to something. Would appreciate help on this one.

File: config/network-policies.yaml [hunk 2]
```diff
@@ -0,0 +1,566 @@
...
+      namespaceSelector:
+        matchLabels:
+          name: #@ data.values.system_namespace
+    - podSelector:
+        matchLabels:
+          app: log-cache
+      namespaceSelector:
+        matchLabels:
+          name: #@ data.values.system_namespace
+  egress:
```
Comment (loewenstein) [inline config/network-policies.yaml:293]: That's deny all, isn't it? This will collide with the [support for external databases](https://github.com/cloudfoundry/cf-for-k8s/issues/168). As the IPs of external databases might not be known or even fixed, I suppose we will have to allow egress traffic to all public IPs, potentially also Cloud Provider network internal IPs.

I would appreciate, if this doesn't get merged without a documented concept of how to open up the egress rules.
At the very least the [external database documentation](https://github.com/cloudfoundry/cf-for-k8s/blob/master/docs/configuration/external-databases.md) should be supported and updated.

Comment (loewenstein) [reply config/network-policies.yaml:293]: Ah ok, RTFM. 
```yaml
egress:
  - {}
```

is actually allow all. As this is work in progress, the general comment about keeping external databases in mind still stands.

Comment (tcdowney) [reply config/network-policies.yaml:293]: Yep, was keeping external databases, external service brokers, external identity providers, etc. in mind for Cloud Controller/UAA.

**Comment by oppegard** [comment]
Hi @tcdowney, here's our (Log Store team) thoughts about this. Note that we're not a core component of CF (e.g. not in cf-deployment or available as an ops-file). We're coming from the viewpoint of a service that is tightly coupled to CF, that a customer may optionally deploy alongside cf-for-k8s. It is deployed in its own namespace, currently `log-store`.

1. We communicate with UAA and cf-api-server to determine if a user is authorized to view certain logs. We currently do this with the internal service names (uaa.cf-system.svc.cluster.local and capi.cf-system.svc.cluster.local). We added `istio-injection: enabled` to our namespace so our pods get the sidecar for handling mTLS to UAA, cf-api-server, and ourselves.

1. We use the cf-system/istio-ingressgateway for external ingress to our service (e.g. log-store.system.my-foundation.com).

1. Presently, we manually configure fluentd to egress logs to a service in our namespace. @pianohacker may have more info on how "syslog sinks" will work in cf-for-k8s.

Things are working well for us currently, but we're not sure what impact NetworkPolicies will have on the above points.

**Comment by tcdowney** [comment]
@oppegard for points 1 and 2 you would need to include `NetworkPolicy` similar to this one for `log-cache`:
https://github.com/cloudfoundry/cf-for-k8s/pull/270/files#diff-652a3e228db9cbb604d4ba5801be220fR484-R529

You could probably remove some of the `ingress` entries (like I don't think `cf-api-server` talks to your component), but that should serve as a decent example.

For the `fluentd` portion.. I think you're just configuring the `fluentd` that is deployed as a `DaemonSet` already right? The policies in this PR right now aren't restricting what `fluentd` can egress to. See:
https://github.com/cloudfoundry/cf-for-k8s/pull/270/files#r449295249

File: docs/component_authors/networking.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+### Namespaces and Istio Sidecar Injection
+CF for Kubernetes uses Istio's [automatic mTLS](https://istio.io/latest/docs/tasks/security/authentication/authn-policy/#auto-mutual-tls) with a `STRICT` authentication policy. This means that all `Pods` that have an Istio sidecar will require any `Pods` that communicate with them to also have an Istio sidecar and use mutual TLS to secure traffic. Traffic that doesn't use mTLS or traffic that does not come from an Istio sidecar will be rejected and the connection will fail.
+
+When you create a new namespace, in order to take advantage of all of Istio's features, and communicate with other `Pods` that Istio sidecars, your namespace must have an `istio-injection=enabled` label to turn on automatic [Sidecar Injection](https://istio.io/latest/docs/setup/additional-setup/sidecar-injection/). This makes it so the pods in your namespace will contain an [Istio sidecar proxy](https://istio.io/latest/docs/reference/config/networking/sidecar/) running alongside them. In CF for Kubernetes, Istio sidecar injection is added automatically by a `ytt` [config](https://github.com/cloudfoundry/cf-for-k8s/blob/master/config/networking.yml#L63).
```
Comment (jamespollard8) [inline docs/component_authors/networking.md:118]: nit: I'd say "..is added automatically by a `ytt` [overlay]

File: config/network-policies.yaml [hunk 4]
```diff
@@ -0,0 +1,563 @@
...
+          cloudfoundry.org/source_type: STG
+      namespaceSelector:
+        matchLabels:
+          name: #@ data.values.staging_namespace
+    - namespaceSelector:
+        matchLabels:
+          cf-for-k8s.cloudfoundry.org/cf-system-ns: ""
+      podSelector:
+        matchLabels:
+          app.kubernetes.io/name: cf-api-server
```
Comment (piyalibanerjee) [inline config/network-policies.yaml:410]: We might need to ad the cf-api-worker to this since it may be used to delete from the blobstore.

File: config/network-policies.yaml [hunk 5]
```diff
@@ -0,0 +1,563 @@
...
+metadata:
+  name: capi-kpack-watcher
+  namespace: #@ data.values.system_namespace
+spec:
+  policyTypes:
+  - Egress
+  podSelector:
+    matchLabels:
+      app.kubernetes.io/name: cf-api-kpack-watcher
+  egress:
```
Comment (cwlbraa) [inline config/network-policies.yaml:464]: right now, to my knowledge, the cf-api-kpack-watcher needs to talk to:

- k8s API
- cf-api-server
- container registry

File: config/network-policies.yaml [hunk 6]
```diff
@@ -0,0 +1,563 @@
...
+      namespaceSelector:
+        matchLabels:
+          cf-for-k8s.cloudfoundry.org/cf-system-ns: ""
+  egress:
+  - {}
+---
+kind: NetworkPolicy
+apiVersion: networking.k8s.io/v1
+metadata:
+  name: capi-kpack-watcher
```
Comment (piyalibanerjee) [inline config/network-policies.yaml:456]: Heads up: this name will change soon after implementing this ADR: https://github.com/cloudfoundry/capi-k8s-release/commit/6863e894c580e44d5652a358fdd33867805de850

**Comment by piyalibanerjee** [comment]
Hello! @cwlbraa and I looked over all the CAPI-related Network Policies. They seem right to us; we left a few comments on some of them.

File: config/network-policies.yaml (hunk 5 above)
Comment (tcdowney) [reply config/network-policies.yaml:464]: This config lets it talk to anything.

Anything that has to talk to the k8s API needs open egress to be generic because the Kubernetes API may be external to the cluster _and_ we don't have a generic way of knowing what IP address(es) it might be at.

That's not to mention the container registry bit.

File: config/network-policies.yaml (hunk 6 above)
Comment (tcdowney) [reply config/network-policies.yaml:456]: The name of the policy doesn't matter, it just needs to make sense to whoever is looking at it. If would like it to be named something differently feel free to update it! 👍

File: docs/component_authors/networking.md [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+When you create a new namespace, in order to take advantage of all of Istio's features, and communicate with other `Pods` that Istio sidecars, your namespace must have an `istio-injection=enabled` label to turn on automatic [Sidecar Injection](https://istio.io/latest/docs/setup/additional-setup/sidecar-injection/). This makes it so the pods in your namespace will contain an [Istio sidecar proxy](https://istio.io/latest/docs/reference/config/networking/sidecar/) running alongside them. In CF for Kubernetes, Istio sidecar injection is added automatically by a `ytt` [overlay](https://github.com/cloudfoundry/cf-for-k8s/blob/master/config/networking.yml#L63).
+
+To confirm your namespace has automatic sidecar injection enabled, you can describe your namespace with `kubectl get namespace <namespace_name> -L istio-injection` and see that Istio Injection is set to "enabled".
+
```
Comment (heycait) [inline docs/component_authors/networking.md:121]: It appears the istio components living in the `istio-system` namespace have a metrics endpoint at ":15014/metrics". However, we intend to deploy Prometheus into the `cf-system` namespace which has the sidecar injection enabled, whereas `istio-system` does not. How would we be able to scrape the istio metrics?

File: config/network-policies.yaml [hunk 8]
```diff
@@ -0,0 +1,563 @@
...
+        matchLabels:
+          cf-for-k8s.cloudfoundry.org/cf-system-ns: ""
+    - podSelector:
+        matchLabels:
+          app.kubernetes.io/name: cf-api-server
+      namespaceSelector:
+        matchLabels:
+          cf-for-k8s.cloudfoundry.org/cf-system-ns: ""
+---
+kind: NetworkPolicy
```
[the remark this answers is not in the document]
Comment (tcdowney) [reply config/network-policies.yaml:528]: You could give a permissive policy to the `prometheus` -- allow it to egress to anything -- and then for the components that have metrics endpoints they could either:

1. Write an ingress policy to explicitly allow prometheus (this would end up in a lot of entries probably) for each component
2. We could coordinate and have system components serve their metrics scrape endpoints on a specific well-known port and have a blanket allow ingress rule for that port
3. Assuming you label prometheus with a well-known label, have a blanket allow ingress rule for the namespace that says that pods should allow ingress from something with that label
3. Some mixture of the above?

File: docs/component_authors/networking.md (hunk 7 above)
Comment (tcdowney) [reply docs/component_authors/networking.md:121]: I don't think it should be a problem. Both Pods must have sidecars if the "server" Pod has an Istio sidecar. The client Pod should be able to talk to other things that don't have one (based on how we've configured it at least).

[client sidecar] ----------any protocol---------> [something *without* sidecar] ✅
[client sidecar] ---------mTLS----------> [something with sidecar] ✅
[client *without* sidecar] ---------any protocol----------> [something with sidecar] 🛑

Comment (tcdowney) [reply docs/component_authors/networking.md:121]: We'll try and take another pass at this section of the docs to clarify this.

Comment (heycait) [reply docs/component_authors/networking.md:121]: So you're saying because istio components don't require it, the Prometheus pod should be able to connect regardless?

I added the network policy below and added annotations to istio-telemetry but am getting this error when Prometheus tries to scrape:  `server returned HTTP status 503 Service Unavailable`.

```
---
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-prometheus-scrape
  namespace: istio-system
spec:
  podSelector: {}
  policyTypes:
  - Ingress
  ingress:
  - from:
    - namespaceSelector:
        matchLabels:
          cf-for-k8s.cloudfoundry.org/cf-system-ns: ""
      podSelector:
        matchLabels:
          what-am-i: prometheus
```

File: config/network-policies.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (heycait) [reply config/network-policies.yaml:528]: Options 2/3 seem like good options with 3 probably being the most ideal. Duplicating ingress across each component would be very unfortunate.

[the remark this answers is not in the document]
Comment (heycait) [reply config/network-policies.yaml:528]: What would Option 3 policy look like?

[the remark this answers is not in the document]
Comment (kkburr) [reply config/network-policies.yaml:528]: I like option 3 as well. Regarding 2, I think CAPI needed to use 9102 for statsd_exporter.

I really like the loop idea and feel like could make implementing a netpol for prometheus simple if we are leaning towards option 1. I might even vote to put prometheus in its own namespace since it doesn't seem like we're gaining anything by putting it in cf-system anymore.

File: docs/component_authors/networking.md (hunk 7 above)
Comment (kkburr) [reply docs/component_authors/networking.md:121]: Do you also have to add a network policy that allows egress from the prometheus pod to that ns?

File: config/network-policies.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (tcdowney) [reply config/network-policies.yaml:528]: Option 3 would look similar to this policy that allows anything in `istio-system` to talk to all of the Pods in `cf-system`:
https://github.com/cloudfoundry/cf-for-k8s/pull/270/files#diff-652a3e228db9cbb604d4ba5801be220fR47

A similar policy could be made that matches to `prometheus`.

[the remark this answers is not in the document]
Comment (tcdowney) [reply config/network-policies.yaml:528]: > I might even vote to put prometheus in its own namespace since it doesn't seem like we're gaining anything by putting it in cf-system anymore.

I think `cf-for-k8s` as a project should have a larger discussion on this. I think there are some operational benefits to consolidating "cf related stuff" in `cf-system`, but am curious to hear others' thoughts here and would love project-wide guidance on the subject. 😅

cc @jenspinney @jamespollard8 @davewalter

[the remark this answers is not in the document]
Comment (jamespollard8) [reply config/network-policies.yaml:528]: I'm inclined to lean into "do what works" here (as opposed to following rules about consistently consolidating or separating components into namespaces).

So, maybe that means put things in cf-system by default AND have no qualms about putting prometheus in its own namespace if that helps us operationally/maintenance-wise.

I'm definitely interested in hearing from others (eg Jen, Dave, etc)

**Comment by shamus** [comment]
Here's the @cloudfoundry/cf-uaa team's 2 cents:

The UAA team doesn't know which components will want to talk to the UAA (whether to retrieve a token, validate it, or take some other action). That is to say, the UAA is used in a general purpose way, like the database. For this reason we don’t know that it makes sense for the UAA to be the shepherd of this NetworkPolicy.

Looking [here](https://github.com/cloudfoundry/cf-for-k8s/blob/master/config/uaa.yml#L110) we see five clients defined, but it isn't clear if those clients are used by the components identified by this networking policy or not.

The policy in this PR is very explicit; it will need to be updated anytime a new component needs access to the UAA.

Another model that would work is a hub / spoke like policy:

```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
 name: allow-uaa-access
spec:
 podSelector:
  matchLabels:
   app: "uaa"
 policyTypes:
 - Ingress
 ingress:
 - from:
  - podSelector:
    matchLabels:
     networking/allow-uaa-access: "true"
```

Which would allow components to opt into being able to reach the UAA.

Which approach is better really depends on the security goals for CF. The benefit of the approach in this PR is its explicitness. The benefit of the hub / spoke approach is its flexibility.

File: config/network-policies.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (heycait) [reply config/network-policies.yaml:528]: > Option 3 would look similar to this policy that allows anything in `istio-system` to talk to all of the Pods in `cf-system`:
> https://github.com/cloudfoundry/cf-for-k8s/pull/270/files#diff-652a3e228db9cbb604d4ba5801be220fR47
> 
> A similar policy could be made that matches to `prometheus`.

I see, and we'd need one of these for all namespaces then. 

I'd also like guidance about what namespaces we should be using as well.

File: docs/component_authors/networking.md (hunk 7 above)
Comment (heycait) [reply docs/component_authors/networking.md:121]: Played around with this more and was actually able to see metrics for some of the istio-components using the above network policy. However only half of the istio components worked (citadel, pilot, sidecar-injector) and the other half fails (galley, telemetry, policy). I can open this as an issue directly on cf-k8s-networking though.

Comment (tcdowney) [reply docs/component_authors/networking.md:121]: Hmm, a lot of those components are being consolidated as part of the Istio 1.6.x upgrade (we'll end up with just two processes `istiod` and `istio-ingressgateway`) so this might change a bit anyway... but I think if you could create the issue that would help us keep it on the radar!

**Comment by tcdowney** [comment]
Thanks for the feedback and reviews, everyone! We acknowledge the friction this adds and are going to discuss it a bit more next week. 

Long way of saying, please don't merge this PR as is right now!

cc @shalako @keshav-pivotal

File: config/network-policies.yaml (hunk 4 above)
Comment (tcdowney) [reply config/network-policies.yaml:410]: @piyalibanerjee made this change:
https://github.com/cloudfoundry/cf-for-k8s/pull/270/commits/1b0f40ef1ba6f7b7fc0b4b4a3d21ed8698930b9b

Added policy for any Cloud Controller component that has a `fog_connection` config.

File: docs/component_authors/networking.md (hunk 7 above)
Comment (tcdowney) [reply docs/component_authors/networking.md]: We attempted to make that part of the docs more clear. Let us know if you have any other suggestions!
https://github.com/cloudfoundry/cf-for-k8s/pull/270/commits/8d3fd6e6d2142f0f58855f5b89d509437d26cde1

File: config/network-policies.yaml (hunk 1 above)
Comment (Benjamintf1) [reply config/network-policies.yaml:581]: @tcdowney at some point(in the not so far future) we plan on making fluentd able to egress to anywhere specified off platform, but I think only to log-cache should be fine at the moment

Comment (Benjamintf1) [reply config/network-policies.yaml:581]: Let me rephrase, I was remembering wrong. So, for fluentd. Fluentd needs to contact(at the moment), the kubernetes api, and also log-cache(fluentd initiates), and will in the future need to contact any user defined source(so contact all makes sense. If there was a way to unlimit external ip's, or for us to dynamically limit ips(might be very difficult and not worth it), maybe we could, but the way I see it unlimited egress from fluentd makes sense to me, then we could hypotheticially limit internal sources.

as far as ingress though, we'll need teams to be able to ingress into fluentd. Obviously unlimited would work, but if we want to limit it further that could work too?

File: config/network-policies.yaml [hunk 9]
```diff
@@ -0,0 +1,581 @@
...
+        matchLabels:
+          cf-for-k8s.cloudfoundry.org/cf-system-ns: ""
+  egress:
+  - to:
+    - podSelector:
+        matchLabels:
+          app.kubernetes.io/name: uaa
+      namespaceSelector:
+        matchLabels:
+          cf-for-k8s.cloudfoundry.org/cf-system-ns: ""
```
Comment (Benjamintf1) [inline config/network-policies.yaml:538]: what's this label and why is it blank?

File: config/network-policies.yaml [hunk 10]
```diff
@@ -0,0 +1,581 @@
...
+  - Ingress
+  - Egress
+  podSelector:
+    matchLabels:
+      app: log-cache
+  ingress:
+  - from:
+    - podSelector:
+        matchLabels:
+          istio: ingressgateway
```
Comment (Benjamintf1) [inline config/network-policies.yaml:515]: Is this what allows it to be accessed from the external lb?

File: config/network-policies.yaml [hunk 11]
```diff
@@ -0,0 +1,581 @@
...
+      namespaceSelector:
+        matchLabels:
+          cf-for-k8s.cloudfoundry.org/istio-system-ns: ""
+    - podSelector:
+        matchLabels:
+          app.kubernetes.io/name: cf-api-server
```
Comment (Benjamintf1) [inline config/network-policies.yaml:521]: What does cf-api-server need to contact log-cache for? I thought the only usecase for this was handled by the metric proxy now, now log-cache? Is there some usecase I'm missing here?

File: config/network-policies.yaml (hunk 9 above)
Comment (tcdowney) [reply config/network-policies.yaml:538]: It's basically a label that's used as a "tag" where you don't really care about it's value. The `cf-system` namespace has this label:
https://github.com/cloudfoundry/cf-for-k8s/blob/a339593d222cec1481a4e2a35be4e6bab0be881a/config/system-namespace.yml#L9

Sso this is basically saying `log-cache` can talk to a `Pod` labeled `app.kubernetes.io/name=uaa` that is in a namespace labeled `cf-for-k8s.cloudfoundry.org/cf-system-ns`.

File: config/network-policies.yaml (hunk 10 above)
Comment (tcdowney) [reply config/network-policies.yaml:515]: Yep. The External lb will point to the `istio-ingressgateway` Pods which will route traffic to log-cache. From loge-cache's perspective it is receiving ingress traffic from a Pod with the `istio=ingressgateway` label in a namespace with the `cf-for-k8s.cloudfoundry.org/istio-system-ns` label.

File: config/network-policies.yaml (hunk 11 above)
Comment (tcdowney) [reply config/network-policies.yaml:521]: I only knew that it used to talk to log-cache in CF for VMs. Maybe it doesn't need to anymore. 🤷 

Looking at the CCNG config it probably doesn't:
https://github.com/cloudfoundry/capi-k8s-release/blob/4a69758bf636c736dfd178b857ec13b46ee398a6/templates/ccng-config.lib.yml#L87-L102

Comment (tcdowney) [reply config/network-policies.yaml:521]: @piyalibanerjee @cwlbraa 

Can you confirm that CC doesn't talk directly to log-cache anymore?

File: config/network-policies.yaml (hunk 10 above)
Comment (Benjamintf1) [reply config/network-policies.yaml:515]: Thanks

**Comment by georgethebeatle** [comment]
Hi all, 

We (Eirini team) added what was missing in the network policies. We validated them on a cluster and all looks good. We have left egress wide open since all our components are using the kube API.

**Comment by tcdowney** [comment]
Thank you @georgethebeatle!

**Comment by tcdowney** [comment]
@cloudfoundry/cf-release-integration 

@keshav-pivotal and @shalako have been talking with some of the other teams and getting alignment on this approach and it sounds like folks are on board now. We described the expected workflow in the [writing a NetworkPolicy section](https://github.com/cloudfoundry/cf-for-k8s/blob/7a9ee9571538c01534661291734c6f510647ff5a/docs/component_authors/networking.md#writing-network-policies) of the docs we added as part of this story.

Anyways, we believe this is ready to merge now! Thanks.

File: config/network-policies.yaml (hunk 11 above)
Comment (davewalter) [reply config/network-policies.yaml:569]: Is this still an open question? If not, could someone please update this thread and resolve it?

File: config/network-policies.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (davewalter) [reply config/network-policies.yaml:594]: Is this still an open question? If not, could someone please update this thread and resolve it?

File: config/network-policies.yaml (hunk 1 above)
Comment (davewalter) [reply config/network-policies.yaml:629]: Is this still an open question? If not, could someone please update this thread and resolve it?

File: config/network-policies.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (tcdowney) [reply config/network-policies.yaml:594]: I think the docs we wrote around [adding a NetworkPolicy](https://github.com/cloudfoundry/cf-for-k8s/blob/7a9ee9571538c01534661291734c6f510647ff5a/docs/component_authors/networking.md#writing-network-policies) for new components addresses this.

File: config/network-policies.yaml (hunk 1 above)
Comment (tcdowney) [reply config/network-policies.yaml:629]: Summarizing what was said:
Since `fluentd` calls out to the Kubernetes API it needs open egress here since we don't know if the API is hosted on cluster or external to the cluster. Additionally it sounds like `fluentd` Open egress here is correct, at least for now.

Since components are sending stuff directly to `fluentd` it sounds like we're missing some allow rules for `ingress` here.

I don't think we want `ingress` to be truly unlimited from all namespaces (I think it would be bad if apps can arbitrarily send stuff to log-cache... might be able to pollute the logs for other apps), so I'll add rules for the `cf-api-*` jobs and `istio-ingressgateway`.. I think those are the things that are or will soon be sending logs directly to `fluentd` via the network.

File: config/network-policies.yaml (hunk 11 above)
Comment (cwlbraa) [reply config/network-policies.yaml:569]: the section of config that tim pointed out is authoritative, afaik.

**Comment by tcdowney** [comment]
@davewalter I moved it into `config/networking`:
https://github.com/cloudfoundry/cf-for-k8s/pull/270/commits/ba622782ee2e6d099633e3bfa47825a14ab34ba6

File: config/network-policies.yaml (hunk 11 above)
Comment (tcdowney) [reply config/network-policies.yaml]: Removed those two `ingress` rules:
https://github.com/cloudfoundry/cf-for-k8s/pull/270/commits/b9b42d035ae86440f5648bfaf4bdfb5355cde6e9

File: config/network-policies.yaml (hunk 1 above)
Comment (tcdowney) [reply config/network-policies.yaml]: Added these `ingress` rules:
https://github.com/cloudfoundry/cf-for-k8s/pull/270/commits/b9b42d035ae86440f5648bfaf4bdfb5355cde6e9
